In [1]:
!pip install pandas sqlalchemy psycopg2
!pip list
!pip install -U sentence-transformers
!pip install psycopg2-binary

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 384.9/384.9 kB 10.5 MB/s eta 0:00:0000:01
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [17 lines of output]
      
      Error: pg_config executable not found.
      
      pg_config is required to build psycopg2 from source.  Please add the directory
      containing pg_config to the $PATH or specify the full executable path with the
      option:
      
          python setup.py build_ext --pg-config /path/to/pg_config build ...
      
      or with the pg_config option in 'setup.cfg'.
      
      If you prefer to avoid building psycopg2 from source, please install the PyPI
      'psycopg2-binary' package instead.
      
      For further information please check the 'doc/src/install.rst' file (also at
      <https://www.psycopg.org/docs/install.html>).
      
      [end of output]
  
  note: This error originates from a subp

### Описание Библиотек и Их Назначение


In [2]:
# Описание Библиотек и Их Назначение

# Обработка и Анализ Данных
import pandas as pd  # Для обработки и анализа данных (таблицы, временные ряды)
import numpy as np   # Для работы с многомерными массивами и математическими операциями

# Работа с Базами Данных
from sqlalchemy.engine.url import URL  # Формирование URL-адресов для подключения к БД
# Определение и взаимодействие с БД
from sqlalchemy import create_engine, MetaData, Table, Column, select,inspect, Integer, String, DECIMAL, CHAR, BIGINT, func, DATE, Float, Text  
from sqlalchemy.exc import ProgrammingError  # Обработка исключений программирования в SQLAlchemy
from sqlalchemy.inspection import inspect  # Инструменты для инспектирования объектов БД
from sqlalchemy.dialects.postgresql import insert  # Операция вставки для PostgreSQL

# Моделирование и Обработка Естественного Языка (NLP)
from sentence_transformers import SentenceTransformer, util, losses, evaluation, InputExample  # Для работы с моделями трансформеров в NLP
from sentence_transformers.util import cos_sim  # Вычисление косинусного сходства
from scipy.spatial.distance import pdist, squareform  # Расстояния между точками в многомерном пространстве

# Машинное Обучение и PyTorch
from sklearn.metrics.pairwise import cosine_similarity  # Вычисление косинусного сходства (scikit-learn)
from torch.utils.data import DataLoader  # Загрузка данных для моделей глубокого обучения

# Регулярные Выражения и Прочее
import re  # Работа с регулярными выражениями

from sqlalchemy import create_engine, inspect, text


/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.24.3
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


# База на сервере

In [3]:
DATABASE = {
    'drivername': 'postgresql',
    'username': 'wasjaip',
    'password': 'qz4bdl63Q1977!',
    'host': '77.222.36.33',
    'port': 19679,
    'database': 'wasjaip',
    'query': {}
}

engine = create_engine(URL.create(**DATABASE))


In [4]:
engine

Engine(postgresql://wasjaip:***@77.222.36.33:19679/wasjaip)

### создание таблицы SQL

In [5]:
# metadata = MetaData(bind=engine)

# # Названия столбцов
# column_names = [
#     'geonameid', 'name', 'asciiname', 'alternatenames', 'latitude', 'longitude',
#     'feature_class', 'feature_code', 'country_code', 'cc2', 'admin1_code',
#     'admin2_code', 'admin3_code', 'admin4_code', 'population', 'elevation',
#     'dem', 'timezone', 'modification_date']

# # Проверка наличия таблицы 'geonames'
# if not inspect(engine).has_table('geonames'):
#     # Создание таблицы, если она не существует
#     geonames = Table('geonames', metadata,
#         Column('geonameid', Integer),
#         Column('name', String(200)),
#         Column('asciiname', String(200)),
#         Column('alternatenames', String(10000)),
#         Column('latitude', DECIMAL),
#         Column('longitude', DECIMAL),
#         Column('feature_class', CHAR(1)),
#         Column('feature_code', String(10)),
#         Column('country_code', CHAR(2)),
#         Column('cc2', String(200)),
#         Column('admin1_code', String(20)),
#         Column('admin2_code', String(80)),
#         Column('admin3_code', String(20)),
#         Column('admin4_code', String(20)),
#         Column('population', BIGINT),
#         Column('elevation', Integer),
#         Column('dem', Integer),
#         Column('timezone', String(40)),
#         Column('modification_date', DATE)
#     )
#     metadata.create_all(engine)

#     # Чтение данных из TSV файла
#     data = pd.read_csv('RU/RU.txt', sep='\t', names=column_names, encoding='utf-8')
#     # Загрузка данных в базу данных
#     data.to_sql('geonames', con=engine, if_exists='append', index=False)
# else:
#     # Если таблица уже существует
#     geonames = Table('geonames', metadata, autoload_with=engine)
#     print("Таблица уже существует.")


In [6]:
# # Имя файла без расширения для названия таблицы
# file_name = 'allCountries'  # Пример, замените на имя вашего файла  allCountries.txt
# table_name = file_name.split('.')[0]
# # Названия столбцов
# column_names = [
#     'geonameid', 'name', 'asciiname', 'alternatenames', 'latitude', 'longitude',
#     'feature_class', 'feature_code', 'country_code', 'cc2', 'admin1_code',
#     'admin2_code', 'admin3_code', 'admin4_code', 'population', 'elevation',
#     'dem', 'timezone', 'modification_date']

# # Проверка наличия таблицы и создание, если она отсутствует
# if not inspect(engine).has_table(table_name):
#     new_table = Table(table_name, metadata,
#         Column('geonameid', Integer),
#         Column('name', String(200)),
#         Column('asciiname', String(200)),
#         Column('alternatenames', String(10000)),
#         Column('latitude', DECIMAL),
#         Column('longitude', DECIMAL),
#         Column('feature_class', CHAR(1)),
#         Column('feature_code', String(10)),
#         Column('country_code', CHAR(2)),
#         Column('cc2', String(200)),
#         Column('admin1_code', String(20)),
#         Column('admin2_code', String(80)),
#         Column('admin3_code', String(20)),
#         Column('admin4_code', String(20)),
#         Column('population', BIGINT),
#         Column('elevation', Integer),
#         Column('dem', Integer),
#         Column('timezone', String(40)),
#         Column('modification_date', DATE)
#     )
#     metadata.create_all(engine)

#     # Чтение данных из файла
#     data = pd.read_csv(f'{file_name}.txt', sep='\t', names=column_names, encoding='utf-8')

#     # Загрузка данных в базу данных
#     data.to_sql(table_name, con=engine, if_exists='replace', index=False)
#     print(f"Таблица '{table_name}' создана и данные загружены.")
# else:
#     print(f"Таблица '{table_name}' уже существует, данные не были загружены.")


### alternateNames

In [7]:
# # Создание таблицы alternateNames, если она не существует
# column_names_al = [
#     'alternateNameId',   # ID альтернативного названия
#     'geonameid',         # ID, ссылающийся на запись в таблице 'geoname'
#     'isolanguage',       # ISO-код языка
#     'alternate_name',    # Альтернативное название или вариант названия
#     'isPreferredName',   # Является ли это официальным/предпочтительным названием
#     'isShortName',       # Является ли это коротким названием
#     'isColloquial',      # Является ли это разговорным или жаргонным термином
#     'isHistoric',        # Является ли это историческим названием
#     'from',              # Период, с которого начинается использование названия
#     'to'                 # Период
# ]
#
#
#
# if not inspect(engine).has_table('alternateNames1'):
#     alternate_names1 = Table('alternateNames1', metadata,
#         Column('alternateNameId', Integer, primary_key=True),
#         Column('geonameid', Integer),
#         Column('isolanguage', String(7)),
#         Column('alternate_name', String(400)),
#         Column('isPreferredName', CHAR(1)),
#         Column('isShortName', CHAR(1)),
#         Column('isColloquial', CHAR(1)),
#         Column('isHistoric', CHAR(1)),
#         Column('from', Integer),
#         Column('to', Integer)
#     )
#     metadata.create_all(engine)
#
#
#     # Чтение данных из файла
#     data = pd.read_csv('alternateNames1.txt', sep='\t', names=column_names_al, encoding='utf-8')
#
#     # Загрузка данных в базу данных
#     data.to_sql(table_name, con=engine, if_exists='replace', index=False)
#     print("Таблица 'alternateNames' создана.")
# else:
#     print("Таблица 'alternateNames' уже существует.")


In [8]:
# # Создание структуры таблицы
# inspector = inspect(engine)
# if not inspector.has_table('alternateNames'):
#     alternate_names = Table('alternateNames', metadata,
#         Column('alternateNameId', Integer, primary_key=True),
#         Column('geonameid', Integer),
#         Column('isolanguage', String(7)),
#         Column('alternate_name', String(400)),
#         Column('isPreferredName', CHAR(1)),
#         Column('isShortName', CHAR(1)),
#         Column('isColloquial', CHAR(1)),
#         Column('isHistoric', CHAR(1)),
#         Column('from', Integer),
#         Column('to', Integer),
#         Column('link', String(255)),  # Дополнительный столбец для ссылок
#         Column('wkdt', String(255)),   # Дополнительный столбец для идентификаторов Wikidata
#         extend_existing=True
#     )
#     metadata.create_all(engine)
#
# # Чтение данных из файла
# data = pd.read_csv('alternateNames.txt', sep='\t', encoding='utf-8', header=None)
# max_columns = 12  # Максимальное ожидаемое количество столбцов
# data = data.reindex(columns=range(max_columns))
# data.columns = ['alternateNameId', 'geonameid', 'isolanguage', 'alternate_name', 'isPreferredName', 'isShortName', 'isColloquial', 'isHistoric', 'from', 'to', 'link', 'wkdt']
#
# # Загрузка данных в базу данных
# data.to_sql('alternateNames', con=engine, if_exists='replace', index=False)
# print("Таблица 'alternateNames' создана.")


все таблицы что получились

In [9]:
engine

Engine(postgresql://wasjaip:***@77.222.36.33:19679/wasjaip)

### Проверка

In [10]:
# Предположим, что engine уже создан
# engine = create_engine('ваш_путь_к_базе_данных')

# Получение списка всех таблиц в базе данных
inspector = inspect(engine)
table_names = inspector.get_table_names()

# Создание соединения для выполнения запросов
with engine.connect() as connection:
    # Перебор таблиц и вывод информации о каждой
    for table_name in table_names:
        # Получение списка столбцов для каждой таблицы
        columns = inspector.get_columns(table_name)
        column_names = [column['name'] for column in columns]

        # Получение количества записей в каждой таблице
        query = text(f'SELECT COUNT(*) FROM "{table_name}"')
        record_count = connection.execute(query).scalar()

        # Вывод информации
        print(f"Таблица: {table_name}")
        print(f"Столбцы: {column_names}")
        print(f"Количество записей: {record_count}")
        print("\n")  # Для разделения информации о разных таблицах

        


Таблица: alternateNames
Столбцы: ['alternateNameId', 'geonameid', 'isolanguage', 'alternate_name', 'isPreferredName', 'isShortName', 'isColloquial', 'isHistoric', 'from', 'to', 'link', 'wkdt']
Количество записей: 16061249


Таблица: allCountries
Столбцы: ['geonameid', 'name', 'asciiname', 'alternatenames', 'latitude', 'longitude', 'feature_class', 'feature_code', 'country_code', 'cc2', 'admin1_code', 'admin2_code', 'admin3_code', 'admin4_code', 'population', 'elevation', 'dem', 'timezone', 'modification_date']
Количество записей: 12493381


Таблица: geonames
Столбцы: ['geonameid', 'name', 'asciiname', 'alternatenames', 'latitude', 'longitude', 'feature_class', 'feature_code', 'country_code', 'cc2', 'admin1_code', 'admin2_code', 'admin3_code', 'admin4_code', 'population', 'elevation', 'dem', 'timezone', 'modification_date']
Количество записей: 374656


Таблица: cities
Столбцы: ['geonameid', 'name', 'asciiname', 'alternatenames', 'latitude', 'longitude', 'feature_class', 'feature_code', 

In [11]:
# # Получение списка всех таблиц в базе данных
# inspector = inspect(engine)
# table_names = inspector.get_table_names()

# # Перебор таблиц и вывод информации о каждой
# for table_name in table_names:
#     # Получение списка столбцов для каждой таблицы
#     columns = inspector.get_columns(table_name)
#     column_names = [column['name'] for column in columns]

#     # Получение количества записей в каждой таблице
#     query = f'SELECT COUNT(*) FROM "{table_name}"'
#     record_count = engine.execute(query).scalar()

#     # Вывод информации
#     print(f"Таблица: {table_name}")
#     print(f"Столбцы: {column_names}")
#     print(f"Количество записей: {record_count}")
#     print("\n")  # Для разделения информации о разных таблицах


Основная таблица 'geoname' имеет следующие поля:
---------------------------------------------------
- `geonameid`: целочисленный идентификатор записи в базе данных geonames.
- `name`: название географического объекта (в кодировке utf8), varchar(200).
- `asciiname`: название географического объекта на английском языке (латинские символы), varchar(200).
- `alternatenames`: альтернативные названия, разделенные запятыми, автоматически транслитерированные ASCII-имена, удобный атрибут из таблицы alternatename, varchar(10000).
- `latitude`: широта в десятичных градусах (WGS84).
- `longitude`: долгота в десятичных градусах (WGS84).
- `feature class`: см. http://www.geonames.org/export/codes.html, char(1).
- `feature code`: см. http://www.geonames.org/export/codes.html, varchar(10).
- `country code`: двухбуквенный код страны по ISO-3166, 2 символа.
- `cc2`: альтернативные коды стран, разделенные запятыми, двухбуквенный код страны по ISO-3166, 200 символов.
- `admin1 code`: код fipscode (может быть изменен на код iso), см. исключения ниже, см. файл admin1Codes.txt для отображения названий этого кода; varchar(20).
- `admin2 code`: код второго уровня административного деления, округ в США, см. файл admin2Codes.txt; varchar(80).
- `admin3 code`: код третьего уровня административного деления, varchar(20).
- `admin4 code`: код четвертого уровня административного деления, varchar(20).
- `population`: bigint (8-байтовый int).
- `elevation`: высота в метрах, integer.
- `dem`: цифровая модель рельефа, srtm3 или gtopo30, средняя высота участка 3''x3'' (примерно 90mx90m) или 30''x30'' (примерно 900mx900m) в метрах, integer. srtm обработан cgiar/ciat.
- `timezone`: идентификатор часового пояса IANA (см. файл timeZone.txt), varchar(40).
- `modification date`: дата последнего изменения в формате yyyy-MM-dd.


Таблица `alternateNames` включает следующие поля:
---------------------------------------------------
- `alternateNameId`: целочисленный идентификатор альтернативного названия.
- `geonameid`: идентификатор, ссылающийся на запись в таблице `geoname`.
- `isolanguage`: ISO-код языка (2 или 3 символа), возможно, с дополнением через дефис и кодом страны для специфических вариантов (например: zh-CN) или названием варианта (например: zh-Hant), 4 символа 'post' для почтовых кодов, 'iata', 'icao' и 'faac' для кодов аэропортов, fr_1793 для названий времен Французской революции, 'abbr' для сокращений, ссылка на веб-сайт (в основном на Wikipedia), 'wkdt' для идентификатора Wikidata, varchar(7).
- `alternate_name`: альтернативное название или вариант названия, varchar(400).
- `isPreferredName`: указывает, является ли это альтернативное название официальным или предпочтительным.
- `isShortName`: указывает, является ли это название коротким (например, "Калифорния" для "Штата Калифорния").
- `isColloquial`: указывает, является ли это альтернативное название разговорным или жаргонным термином (например, "Большое Яблоко" для "Нью-Йорка").
- `isHistoric`: указывает, является ли это альтернативное название историческим и использовалось в прошлом (например, "Бомбей" для "Мумбаи").
- `from`: период, с которого начинается использование названия.
- `to`: период, до которого использовалось название.


In [12]:
"все ок"

'все ок'

. RU, BY, KG, KZ, AM, GE, RS, ME

мне нужно из всех таблиц собрать df   :  'name', 'asciiname', 'alternatenames' 'country_code'

In [13]:
# Определение запроса для основных таблиц
query_main = """
SELECT geonameid, name, asciiname, country_code, admin1_code, admin2_code
FROM {}
WHERE country_code IN ('RU', 'BY', 'KG', 'KZ', 'AM', 'GE', 'RS', 'ME')
"""

Таблица: alternateNames
Столбцы: ['alternateNameId', 'geonameid', 'isolanguage', 'alternate_name', 'isPreferredName', 'isShortName', 'isColloquial', 'isHistoric', 'from', 'to', 'link', 'wkdt']
Количество записей: 16061249

In [14]:
# Запрос для таблицы alternateNames
query_alt_names = """
SELECT geonameid, alternate_name, "isPreferredName", "isShortName", "isColloquial", "isHistoric", "to"
FROM "alternateNames"
"""



In [15]:
# Чтение данных из таблиц
df_main = pd.concat([
    pd.read_sql_query(query_main.format(table), engine)
    for table in ['"allCountries"', '"cities"', '"cities15000"', '"cities500"', '"cities5000"', '"geonames"']
], ignore_index=True)

In [16]:
df_alt_names = pd.read_sql_query(query_alt_names, engine)

In [17]:

# Объединение данных
df_combined = pd.merge(df_main, df_alt_names, on='geonameid', how='left')

# Вывод первых нескольких строк для проверки
df_combined.head(100)


geonameid             name        asciiname country_code admin1_code  \
0     616075         Voskehat         Voskehat           AM          01   
1     616075         Voskehat         Voskehat           AM          01   
2     616075         Voskehat         Voskehat           AM          01   
3     616075         Voskehat         Voskehat           AM          01   
4     616075         Voskehat         Voskehat           AM          01   
..       ...              ...              ...          ...         ...   
95    616094   Verin Getashen   Verin Getashen           AM          04   
96    616095  Verin Dzhrashen  Verin Dzhrashen           AM          11   
97    616095  Verin Dzhrashen  Verin Dzhrashen           AM          11   
98    616095  Verin Dzhrashen  Verin Dzhrashen           AM          11   
99    616096    Verin Dzhrapi    Verin Dzhrapi           AM          00   

   admin2_code                                alternate_name  isPreferredName  \
0         None                                     Patrindzh              NaN   
1         None                                      Patridzh              NaN   
2         None                                       Voskeat              NaN   
3         None                                      Воскеваз              NaN   
4         None                                      Voskehat              NaN   
..         ...                                           ...              ...   
95        None  https://en.wikipedia.org/wiki/Verin_Getashen              NaN   
96        None                                          Tolk              NaN   
97        None                               Verin Dzhrashen              NaN   
98        None                                      Q7921350              NaN   
99        None                                        Kegach              NaN   

    isShortName  isColloquial  isHistoric    to  
0           NaN           NaN         NaN  None  
1           NaN           NaN         NaN  None  
2           NaN           NaN         NaN  None  
3           NaN           NaN         NaN  None  
4           NaN           NaN         NaN  None  
..          ...           ...         ...   ...  
95          NaN           NaN         NaN  None  
96          NaN           NaN         NaN  None  
97          NaN           NaN         NaN  None  
98          NaN           NaN         NaN  None  
99          NaN           NaN         NaN  None  

[100 rows x 12 columns]

In [18]:
len(df_combined)

2850723

In [19]:
df_combined.columns

Index(['geonameid', 'name', 'asciiname', 'country_code', 'admin1_code',
       'admin2_code', 'alternate_name', 'isPreferredName', 'isShortName',
       'isColloquial', 'isHistoric', 'to'],
      dtype='object')

In [20]:
df_combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2850723 entries, 0 to 2850722
Data columns (total 12 columns):
 #   Column           Dtype  
---  ------           -----  
 0   geonameid        object 
 1   name             object 
 2   asciiname        object 
 3   country_code     object 
 4   admin1_code      object 
 5   admin2_code      object 
 6   alternate_name   object 
 7   isPreferredName  float64
 8   isShortName      float64
 9   isColloquial     float64
 10  isHistoric       float64
 11  to               object 
dtypes: float64(4), object(8)
memory usage: 261.0+ MB


In [21]:
# Преобразование всех значений в столбце alternate_name в строки
df_combined['alternate_name'] = df_combined['alternate_name'].apply(lambda x: ','.join(map(str, x)) if isinstance(x, list) else str(x))

# Вывод первых нескольких строк для проверки
df_combined[['asciiname', 'alternate_name']].head()



asciiname alternate_name
0  Voskehat      Patrindzh
1  Voskehat       Patridzh
2  Voskehat        Voskeat
3  Voskehat       Воскеваз
4  Voskehat       Voskehat

In [22]:

# Подсчет количества значений в столбце alternate_name, содержащих 'https://'
count_links = df_combined['alternate_name'].str.contains('https://').sum()

print(f"Количество значений со ссылками 'https://': {count_links}")


Количество значений со ссылками 'https://': 44652


In [23]:
# Удаление значений в столбце alternate_name, содержащих 'https://'
df_combined['alternate_name'] = df_combined['alternate_name'].apply(lambda x: "" if 'https://' in x else x)

# Вывод первых нескольких строк для провер


In [24]:
# Вывод первых нескольких строк для проверки
df_combined[['asciiname', 'alternate_name']].head()

asciiname alternate_name
0  Voskehat      Patrindzh
1  Voskehat       Patridzh
2  Voskehat        Voskeat
3  Voskehat       Воскеваз
4  Voskehat       Voskehat

In [25]:
# Поиск и отображение всех строк, где alternate_name является NaN
nan_rows = df_combined[df_combined['alternate_name'].isna()]

# Вывод строк с NaN в столбце alternate_name
nan_rows

Empty DataFrame
Columns: [geonameid, name, asciiname, country_code, admin1_code, admin2_code, alternate_name, isPreferredName, isShortName, isColloquial, isHistoric, to]
Index: []

In [26]:

# Загрузка вашего DataFrame
# Пример: df_combined = pd.read_csv('path_to_your_dataframe.csv')

# # Фильтрация ненужных столбцов
# filtered_df = df_combined[['asciiname', 'country_code', 'admin1_code', 'admin2_code', 'alternate_name']]

# Фильтрация ненужных столбцов
filtered_df = df_combined[['asciiname', 'country_code', 'admin1_code', 'alternate_name']]

# Группировка данных по asciiname, country_code, admin1_code, admin2_code и агрегация всех alternate_name
grouped_df = filtered_df.groupby(['asciiname', 'country_code', 'admin1_code'])['alternate_name'].apply(list).reset_index()



In [27]:
# Вывод первых нескольких строк сгруппированных данных для проверки
grouped_df.head(1000)

asciiname country_code admin1_code  \
0                                 (Brickworks)           RS          SE   
1                                (Lumber mill)           RS          SE   
2    (former) Roman Catholic Diocese of Doclea           ME        16.0   
3                                      1448 Km           RU        49.0   
4                                      17-Y km           RU        64.0   
..                                         ...          ...         ...   
995                                 Adamovichi           BY         3.0   
996                                 Adamovichi           BY         5.0   
997                                   Adamovka           BY         2.0   
998                                   Adamovka           BY         7.0   
999                                   Adamovka           RU        25.0   

                                        alternate_name  
0                                       [(Brickworks)]  
1                                      [(Lumber mill)]  
2                                                [nan]  
3                                           [nan, nan]  
4                               [17-й км, , 17-й км, ]  
..                                                 ...  
995    [Adamowicze, Adamoviche, Adamovichi, Адамовичи]  
996                            [Adamovichi, Адамовичи]  
997                               [Adamovka, Адамовка]  
998           [Adamovka, Адамовка, Adamovka, Адамовка]  
999  [Damovka, Адамовка, Adamovka, Adamovka, Damovk...  

[1000 rows x 4 columns]

In [28]:
grouped_df.shape

(551316, 4)

In [29]:
# Удаление дубликатов в списках столбца alternate_name
grouped_df['alternate_name'] = grouped_df['alternate_name'].apply(lambda x: list(set(x)))

# Вывод первых нескольких строк для проверки
grouped_df.head(100)


asciiname country_code admin1_code  \
0                                (Brickworks)           RS          SE   
1                               (Lumber mill)           RS          SE   
2   (former) Roman Catholic Diocese of Doclea           ME        16.0   
3                                     1448 Km           RU        49.0   
4                                     17-Y km           RU        64.0   
..                                        ...          ...         ...   
95                                   Ababkovo           RU          60   
96                                   Ababkovo           RU        77.0   
97                                  Ababurovo           RU          37   
98                                  Ababurovo           RU        37.0   
99                                  Ababurovo           RU          47   

            alternate_name  
0           [(Brickworks)]  
1          [(Lumber mill)]  
2                    [nan]  
3                    [nan]  
4              [, 17-й км]  
..                     ...  
95    [Ababkovo, Абабково]  
96    [Ababkovo, Абабково]  
97  [Абабурово, Ababurovo]  
98  [Абабурово, Ababurovo]  
99                   [nan]  

[100 rows x 4 columns]

In [30]:
# Загрузка вашего DataFrame
# Пример: grouped_df = pd.read_csv('path_to_your_dataframe.csv')

# Функция для проверки, содержит ли список только NaN
def contains_only_nan(lst):
    return all(pd.isna(item) for item in lst)

# Подсчет количества строк, где alternate_name содержит только NaN
count_nan_only = grouped_df['alternate_name'].apply(contains_only_nan).sum()

print(f"Количество строк, где 'alternate_name' содержит только [nan]: {count_nan_only}")



Количество строк, где 'alternate_name' содержит только [nan]: 0


In [31]:
grouped_df.head(1000)

asciiname country_code admin1_code  \
0                                 (Brickworks)           RS          SE   
1                                (Lumber mill)           RS          SE   
2    (former) Roman Catholic Diocese of Doclea           ME        16.0   
3                                      1448 Km           RU        49.0   
4                                      17-Y km           RU        64.0   
..                                         ...          ...         ...   
995                                 Adamovichi           BY         3.0   
996                                 Adamovichi           BY         5.0   
997                                   Adamovka           BY         2.0   
998                                   Adamovka           BY         7.0   
999                                   Adamovka           RU        25.0   

                                      alternate_name  
0                                     [(Brickworks)]  
1                                    [(Lumber mill)]  
2                                              [nan]  
3                                              [nan]  
4                                        [, 17-й км]  
..                                               ...  
995  [Adamowicze, Адамовичи, Adamovichi, Adamoviche]  
996                          [Adamovichi, Адамовичи]  
997                             [Adamovka, Адамовка]  
998                             [Adamovka, Адамовка]  
999                    [Damovka, Adamovka, Адамовка]  

[1000 rows x 4 columns]

In [32]:
# Функция для замены [nan] значением из asciiname
# Функция для замены списка, содержащего только NaN или пустые строки, значением из asciiname
def replace_empty_with_asciiname(alternate_name, asciiname):
    # Проверка, содержит ли список только NaN или пустые строки
    if all(pd.isna(name) or name == '' for name in alternate_name):
        return [asciiname]
    return alternate_name

# Применение функции к DataFrame
grouped_df['alternate_name'] = grouped_df.apply(
    lambda row: replace_empty_with_asciiname(row['alternate_name'], row['asciiname']),
    axis=1
)


# Вывод первых нескольких строк для проверки
grouped_df.head(1000)

asciiname country_code admin1_code  \
0                                 (Brickworks)           RS          SE   
1                                (Lumber mill)           RS          SE   
2    (former) Roman Catholic Diocese of Doclea           ME        16.0   
3                                      1448 Km           RU        49.0   
4                                      17-Y km           RU        64.0   
..                                         ...          ...         ...   
995                                 Adamovichi           BY         3.0   
996                                 Adamovichi           BY         5.0   
997                                   Adamovka           BY         2.0   
998                                   Adamovka           BY         7.0   
999                                   Adamovka           RU        25.0   

                                      alternate_name  
0                                     [(Brickworks)]  
1                                    [(Lumber mill)]  
2                                              [nan]  
3                                              [nan]  
4                                        [, 17-й км]  
..                                               ...  
995  [Adamowicze, Адамовичи, Adamovichi, Adamoviche]  
996                          [Adamovichi, Адамовичи]  
997                             [Adamovka, Адамовка]  
998                             [Adamovka, Адамовка]  
999                    [Damovka, Adamovka, Адамовка]  

[1000 rows x 4 columns]

In [33]:
# Загрузка вашего DataFrame
# Пример: grouped_df = pd.read_csv('path_to_your_dataframe.csv')

# Функция для замены списков с 'nan' на значение из asciiname
def replace_nan_list_with_asciiname(alternate_name, asciiname):
    # Проверка на наличие 'nan' в списке
    if 'nan' in alternate_name or len(alternate_name) == 0:
        return [asciiname]
    return alternate_name

# Применение функции к DataFrame
grouped_df['alternate_name'] = grouped_df.apply(
    lambda row: replace_nan_list_with_asciiname(row['alternate_name'], row['asciiname']),
    axis=1
)


In [34]:
# Проверка результатов
grouped_df[['asciiname', 'alternate_name']].head(1000)

asciiname  \
0                                 (Brickworks)   
1                                (Lumber mill)   
2    (former) Roman Catholic Diocese of Doclea   
3                                      1448 Km   
4                                      17-Y km   
..                                         ...   
995                                 Adamovichi   
996                                 Adamovichi   
997                                   Adamovka   
998                                   Adamovka   
999                                   Adamovka   

                                      alternate_name  
0                                     [(Brickworks)]  
1                                    [(Lumber mill)]  
2        [(former) Roman Catholic Diocese of Doclea]  
3                                          [1448 Km]  
4                                        [, 17-й км]  
..                                               ...  
995  [Adamowicze, Адамовичи, Adamovichi, Adamoviche]  
996                          [Adamovichi, Адамовичи]  
997                             [Adamovka, Адамовка]  
998                             [Adamovka, Адамовка]  
999                    [Damovka, Adamovka, Адамовка]  

[1000 rows x 2 columns]

In [35]:
grouped_df.shape

(551316, 4)

In [36]:
grouped_df.head(1000)

asciiname country_code admin1_code  \
0                                 (Brickworks)           RS          SE   
1                                (Lumber mill)           RS          SE   
2    (former) Roman Catholic Diocese of Doclea           ME        16.0   
3                                      1448 Km           RU        49.0   
4                                      17-Y km           RU        64.0   
..                                         ...          ...         ...   
995                                 Adamovichi           BY         3.0   
996                                 Adamovichi           BY         5.0   
997                                   Adamovka           BY         2.0   
998                                   Adamovka           BY         7.0   
999                                   Adamovka           RU        25.0   

                                      alternate_name  
0                                     [(Brickworks)]  
1                                    [(Lumber mill)]  
2        [(former) Roman Catholic Diocese of Doclea]  
3                                          [1448 Km]  
4                                        [, 17-й км]  
..                                               ...  
995  [Adamowicze, Адамовичи, Adamovichi, Adamoviche]  
996                          [Adamovichi, Адамовичи]  
997                             [Adamovka, Адамовка]  
998                             [Adamovka, Адамовка]  
999                    [Damovka, Adamovka, Адамовка]  

[1000 rows x 4 columns]

для теста потом убрать

In [37]:
grouped_df1=grouped_df.copy()

In [38]:
grouped_df1.head()

asciiname country_code admin1_code  \
0                               (Brickworks)           RS          SE   
1                              (Lumber mill)           RS          SE   
2  (former) Roman Catholic Diocese of Doclea           ME        16.0   
3                                    1448 Km           RU        49.0   
4                                    17-Y km           RU        64.0   

                                alternate_name  
0                               [(Brickworks)]  
1                              [(Lumber mill)]  
2  [(former) Roman Catholic Diocese of Doclea]  
3                                    [1448 Km]  
4                                  [, 17-й км]

In [39]:
print(grouped_df1['alternate_name'].apply(type).value_counts())
print(grouped_df1['alternate_name'].head())

alternate_name
<class 'list'>    551316
Name: count, dtype: int64
0                                 [(Brickworks)]
1                                [(Lumber mill)]
2    [(former) Roman Catholic Diocese of Doclea]
3                                      [1448 Km]
4                                    [, 17-й км]
Name: alternate_name, dtype: object


In [40]:
# Применяем explode к столбцу 'alternate_name'
grouped_df1 = grouped_df1.explode('alternate_name')

# Фильтруем DataFrame, чтобы исключить строки, где 'asciiname' равен 'alternate_name'
grouped_df1 = grouped_df1[grouped_df1.asciiname != grouped_df1.alternate_name]



In [41]:
grouped_df1=grouped_df1.drop_duplicates(subset=['asciiname','alternate_name'])

In [42]:
# Удаляем строки, где alternate_name - это NaN
grouped_df1 = grouped_df1.dropna(subset=['alternate_name'])

# Удаляем строки, где alternate_name содержит только пробелы
grouped_df1 = grouped_df1[grouped_df1['alternate_name'].str.strip().str.len() > 0]



In [43]:
grouped_df1.shape

(618256, 4)

In [44]:
grouped_df1

asciiname country_code admin1_code  \
4                                 17-Y km           RU        64.0   
5                            189 Kilometr           KZ    12510145   
6        1917t'. Heghap'okhut'yan Zoherin           AM         3.0   
6        1917t'. Heghap'okhut'yan Zoherin           AM         3.0   
7       1930-1940t't'. Brrnut'yan Zoherin           AM         2.0   
...                                   ...          ...         ...   
551287                  oz. Zsigilevskoye           RU        42.0   
551289                   platform Voronki           RU           8   
551289                   platform Voronki           RU           8   
551289                   platform Voronki           RU           8   
551315                  zaliv Sjarunlahti           RU        42.0   

                          alternate_name  
4                                17-й км  
5                           189 Километр  
6       1917t’. Heghap’okhut’yan Zoherin  
6           1917թ. հեղափոխության զոհերին  
7         1930-1940թթ. բռնության զոհերին  
...                                  ...  
551287                        Teinijärvi  
551289                           Voronki  
551289                 платформа Воронки  
551289                 Stantsiya Voronki  
551315                        Särynlahti  

[618256 rows x 4 columns]

In [45]:
grouped_df1['exp'] = grouped_df1.apply(lambda x: InputExample(texts=[x['asciiname'], x['alternate_name']]), axis=1)


In [46]:
grouped_df1['exp']

4          <InputExample> label: 0, texts: 17-Y km; 17-й км
5         <InputExample> label: 0, texts: 189 Kilometr; ...
6         <InputExample> label: 0, texts: 1917t'. Heghap...
6         <InputExample> label: 0, texts: 1917t'. Heghap...
7         <InputExample> label: 0, texts: 1930-1940t't'....
                                ...                        
551287    <InputExample> label: 0, texts: oz. Zsigilevsk...
551289    <InputExample> label: 0, texts: platform Voron...
551289    <InputExample> label: 0, texts: platform Voron...
551289    <InputExample> label: 0, texts: platform Voron...
551315    <InputExample> label: 0, texts: zaliv Sjarunla...
Name: exp, Length: 618256, dtype: object

In [47]:
train_df = grouped_df1['exp'].to_list()

## потестируем разные модели

In [48]:
model = SentenceTransformer('LaBSE')
#model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')


.gitattributes:   0%|          | 0.00/391 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

2_Dense/config.json:   0%|          | 0.00/114 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/2.36M [00:00<?, ?B/s]

README.md:   0%|          | 0.00/2.19k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/804 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.88G [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.62M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/411 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/5.22M [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/461 [00:00<?, ?B/s]

In [49]:
model

SentenceTransformer(
  (0): Transformer({'max_seq_length': 256, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': True, 'pooling_mode_mean_tokens': False, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False})
  (2): Dense({'in_features': 768, 'out_features': 768, 'bias': True, 'activation_function': 'torch.nn.modules.activation.Tanh'})
  (3): Normalize()
)


- Перевод модели в 16-битный формат (через метод .half() в PyTorch) действительно уменьшает использование памяти на уровне каждого параметра, но количество параметров остается неизменным. Поэтому, когда вы считаете общее количество параметров, оно останется таким же, как и в 32-битном формате.

- Однако, каждый параметр теперь занимает в два раза меньше памяти (2 байта вместо 4 байтов), потому что torch.float16 использует 16 бит (2 байта) вместо 32 бит (4 байта) для каждого числа с плавающей точкой. Следовательно, общий объем занимаемой памяти уменьшится вдвое.

Однако, важно понимать, что квантизация модели в PyTorch обычно используется для оптимизации скорости инференса, а не для обучения моделей. В большинстве случаев, квантизация применяется к уже обученным моделям, чтобы ускорить их работу на устройствах с ограниченными ресурсами (особенно на CPU), и может не быть совместима с процессом обучения.

In [50]:

# model = model.half()

In [51]:
# model

In [52]:
# Создание DataLoader
train_dataloader = DataLoader(train_df, shuffle=True, batch_size=64)

Выбор между косинусной потерей (CosineSimilarityLoss) и потерей MultipleNegativesRankingLoss зависит от конкретной задачи и данных, с которыми я работаю. Вот некоторые рассуждения, которые помогают мне сделать выбор:

### CosineSimilarityLoss 😊

- **Используется для**: Этот тип потерь часто использую в задачах, где важно измерить степень схожести или различия между парами предложений 🤔. Он хорошо работает, когда у меня есть чётко определенные положительные и отрицательные пары 👍👎.
- **Принцип работы**: Основывается на косинусной мере схожести, которая измеряет косинус угла между двумя векторами 📐. Этот метод стремится минимизировать угол между парами схожих предложений и максимизировать угол между непохожими 🔍.
- **Подходит для**: Задач, где я хочу, чтобы модель училась различать, насколько близки или далеки тексты по смыслу 📖.

### MultipleNegativesRankingLoss 🤨

- **Используется для**: Этот тип потерь использую в сценариях, когда каждый пример в пакете обучения рассматривается как положительный пример, а все остальные примеры в пакете являются отрицательными. Это хорошо подходит для задач, где нет чётких пар положительных/отрицательных примеров 🚫.
- **Принцип работы**: Учит модель различать положительный пример от набора отрицательных примеров в пакете 🌟. Это делается путём увеличения расстояния между положительными и всеми отрицательными примерами 📏.
- **Подходит для**: Задач ранжирования или рекомендации, где важно выделить наиболее релевантные элементы из набора возможных вариантов 🎯.

### Ваш выбор 🤔

- **Понимание моих данных**: Если у меня есть чётко определённые пары предложений (например, переводы или пары вопрос-ответ), CosineSimilarityLoss может быть более подходящим 👌. Он поможет моей модели научиться различать, насколько близки тексты по смыслу 💬.
- **Цель моделирования**: Если моя цель — обучить модель определять релевантность предложения среди множества других предложений (например, в задачах поиска или рекомендации), MultipleNegativesRankingLoss может быть более подходящим выбором 🔍.

### Итог 🌟

Мой выбор должен базироваться на специфике моих данных и конечной цели модели. Если возможно, я могу провести эксперименты с обоими вариантами потерь и сравнить результаты на валидационном наборе данных, чтобы определить, какой из них лучше работает в моём конкретном случае 🧪📊.



In [53]:
# Использование  потери  2 варинта 
# train_loss = losses.CosineSimilarityLoss(model)
train_loss = losses.MultipleNegativesRankingLoss(model)

При настройке обучения моей модели с помощью `sentence_transformers`, я сталкиваюсь с различными вариантами выбора параметров. Вот как я рассматриваю каждый из них:

### Основные Параметры Обучения 😎

- **train_objectives**: Это комбинация загрузчика данных (`train_dataloader`) и функции потерь (`train_loss`). Это основа моего процесса обучения, определяющая, какие данные используются и как измеряется их эффективность.
- **epochs**: Я задаю количество эпох, в течение которых модель будет обучаться. Эпоха – это один полный проход по всему обучающему набору данных. Например, `epochs=8` означает, что весь набор данных будет пройден восемь раз.

### Расширенные Настройки 🧐

- **warmup_steps (необязательно)**: Это период разогрева, который помогает предотвратить слишком резкие изменения весов в начале обучения. Если я не укажу этот параметр, процесс обучения может быть менее стабильным.
- **evaluator (необязательно)**: Оценщик позволяет мне следить за прогрессом и эффективностью моей модели на валидационных данных во время обучения. Если я не использую оценщик, я не смогу наблюдать за прогрессом до окончания обучения.
- **evaluation_steps (необязательно)**: Этот параметр определяет, как часто я буду оценивать модель во время обучения. Он работает в тандеме с оценщиком.

### Варианты Использования 🤔

- **Быстрое Прототипирование**: Если моя цель – быстро протестировать модель, я могу использовать только основные параметры обучения.
- **Тщательная Настройка и Мониторинг**: Для более точного контроля над процессом обучения и для получения обратной связи в процессе, я выбираю расширенные настройки с оценщиком и шагами разогрева.

### Мой Выбор 🌟

В зависимости от моих конкретных нужд и ограничений проекта, я могу выбрать различные стратегии обучения. Важно помнить, что каждая опция имеет свои преимущества и компромиссы. Экспериментирование и понимание моих данных и целей помогают мне сделать наилучший выбор! 🚀


In [54]:
# Определение оценщика и параметров обучения
evaluator = evaluation.EmbeddingSimilarityEvaluator.from_input_examples(train_df, name='asciiname-evaluator')
model.fit(train_objectives=[(train_dataloader, train_loss)], epochs=6, warmup_steps=100, evaluator=evaluator, evaluation_steps=5000)
# model.fit(train_objectives=[(train_dataloader, train_loss)], epochs=1)

Epoch:   0%|          | 0/6 [00:00<?, ?it/s]

Iteration:   0%|          | 0/9661 [00:00<?, ?it/s]

/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())


Iteration:   0%|          | 0/9661 [00:00<?, ?it/s]

/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())


Iteration:   0%|          | 0/9661 [00:00<?, ?it/s]

/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())


Iteration:   0%|          | 0/9661 [00:00<?, ?it/s]

/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())


Iteration:   0%|          | 0/9661 [00:00<?, ?it/s]

/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())


Iteration:   0%|          | 0/9661 [00:00<?, ?it/s]

/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(PearsonRConstantInputWarning())
/opt/conda/lib/python3.10/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())


In [55]:
model

SentenceTransformer(
  (0): Transformer({'max_seq_length': 256, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': True, 'pooling_mode_mean_tokens': False, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False})
  (2): Dense({'in_features': 768, 'out_features': 768, 'bias': True, 'activation_function': 'torch.nn.modules.activation.Tanh'})
  (3): Normalize()
)

In [56]:
#model.fit(train_objectives=[(train_dataloader, train_loss)], epochs=8)

In [57]:
# ##запишем модель
# !pip install diffusers
# !pip install safetensors


In [58]:
# !pip show safetensors

In [59]:
# !pip show diffusers

In [60]:
import os

In [61]:
# Сохранение модели в директории, доступной на Kaggle
model.save('/kaggle/working/model')

In [62]:
# Проверка файлов в директории сохранения
print(os.listdir('/kaggle/working/model'))

['1_Pooling', 'sentence_bert_config.json', 'README.md', '3_Normalize', 'config_sentence_transformers.json', 'model.safetensors', '2_Dense', 'vocab.txt', 'config.json', 'tokenizer.json', 'modules.json', 'tokenizer_config.json', 'special_tokens_map.json']


In [63]:
!tar -czf model.tar.gz -C /kaggle/working/model .

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


In [64]:
os.path.exists('/kaggle/working/model.tar.gz')

True

проверка 

In [65]:
name=grouped_df1.asciiname.drop_duplicates().values
name[:20]

array(['17-Y km', '189 Kilometr', "1917t'. Heghap'okhut'yan Zoherin",
       "1930-1940t't'. Brrnut'yan Zoherin", '1937 Brrnadatvatsnerin',
       "1937-49t't'. Brrnadatvatsneri", '1937-i Tsnvatsneri',
       "1937t'. Ev Arts'akhyan Paterazmum Zohvatsneri Yerkrord Ashkharhamartum",
       "1990t'. Erdman Araroghut'yan Masnakits'neri", '206 Km', '24 km',
       '26 Komisarneri', '2603 Ostanovochnyy Punkt', '2750-amyakin',
       '2750-amyakin Hushardzan', '2912 Km', '2926 Km', '2966 Km',
       '2997 Km Moshkar', '3017 Km'], dtype=object)

In [66]:
embeddings=model.encode(name)
embeddings.shape

Batches:   0%|          | 0/9705 [00:00<?, ?it/s]

(310545, 768)

In [67]:
embeddings

array([[-0.03105616,  0.0556125 ,  0.04436076, ...,  0.00608696,
         0.0121627 , -0.02921132],
       [ 0.0157514 ,  0.03609248,  0.02846927, ..., -0.06078377,
        -0.03792637,  0.00754857],
       [ 0.00935533,  0.0215654 ,  0.02193071, ..., -0.01009489,
        -0.00157078, -0.041862  ],
       ...,
       [ 0.01569463,  0.03590047, -0.00636198, ..., -0.02944997,
         0.00190718,  0.05633597],
       [-0.00527813,  0.03547215, -0.02337411, ..., -0.01547332,
        -0.02895672,  0.02282221],
       [-0.00900305, -0.00622461,  0.04587244, ..., -0.01499317,
         0.06658256,  0.01263921]], dtype=float32)

In [68]:
# def get_all(geo_name, model=model, embeddings=embeddings, name=name, top=3):
#     # Кодирование запроса в эмбеддинг
#     query_embedding = model.encode([geo_name])

#     # Поиск top наиболее близких эмбеддингов
#     search_results = util.semantic_search(query_embedding, embeddings, top=top)[0]

#     # Преобразование результатов в DataFrame
#     result = pd.DataFrame(search_results)

#     # Добавление имени в результаты на основе corpus_id
#     result = result.assign(name=[name[i] for i in result['corpus_id']])
    
#     return result


In [69]:
# def get_all(geo_name, model=model, embeddings=embeddings, name=name, top=3):
#     # Кодирование запроса в эмбеддинг
#     query_embedding = model.encode([geo_name])

#     # Поиск top наиболее близких эмбеддингов
#     search_results = util.semantic_search(query_embedding, embeddings, top)[0]

#     # Преобразование результатов в DataFrame
#     result = pd.DataFrame(search_results)

#     # Добавление имени в результаты на основе corpus_id
#     result = result.assign(name=[name[i] for i in result['corpus_id']])
    
#     return result


In [72]:
def get_all(geo_name, model=model, embeddings=embeddings, name=name, top=3):
    # Кодирование запроса в эмбеддинг
    query_embedding = model.encode([geo_name])

    # Поиск наиболее близких эмбеддингов
    search_results = util.semantic_search(query_embedding, embeddings, top_k=top)[0]

    # Ограничение результатов до заданного количества top
    search_results = search_results[:top]

    # Преобразование результатов в DataFrame
    result = pd.DataFrame(search_results)

    # Добавление имени в результаты на основе corpus_id
    result = result.assign(name=[name[i] for i in result['corpus_id']])

    return result

In [78]:
# def get_all(geo_name,name=name,embeddings=embeddings,model=model,top=3):
#     result=pd.DataFrame(util.semantic_search(model.encode(geo_name),embeddings,top=top)[0])
#     return result.assing(name=name[result.corpus_id])
get_all('СПб')


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

corpus_id     score              name
0     221595  0.668791  Saint Petersburg
1     222880  0.646645   Sankt-Peterburg
2     239369  0.622244            Spassk

	corpus_id	score	name
0	97513	0.764922	Kirov
1	97426	0.706236	Kirikov
2	97526	0.696088	Kirovsk

	corpus_id	score	name
0	97426	0.799164	Kirikov
1	97513	0.765754	Kirov
2	97420	0.723782	K
